### PR TITLE
feat(system): Filter system metadata by workspace and variable chaining

### PR DESCRIPTION
**Pull request by cameronwaterman**
# Pull Request

## 🤨 Rationale

https://ni.visualstudio.com/DevCentral/_workitems/edit/2527391
https://ni.visualstudio.com/DevCentral/_workitems/edit/2541892

## 👩‍💻 Implementation

- Add workspace field to metadata query which is capable of resolving variables.
- Add variable resolution logic to `metricFindQuery`.

## 🧪 Testing

Added tests and existing tests continue to pass.

## ✅ Checklist

- [X] This PR has a title that follows the [commit message format](https://github.com/ni/systemlink-grafana-plugins#commit-message-format).

### CODE REVIEW

File: src/datasources/system/components/SystemQueryEditor.tsx [hunk 1]
```diff
@@ -49,4 +69,5 @@ const tooltips = {
               Summary allows you to visualize the number of disconnected and connected systems.`,
   system: `Query for a specific system by its name or ID. If left blank, the plugin returns all
             available systems. You can enter a variable into this field.`,
+  workspace: `The workspace to search for the given system.`,
```
Comment (cameronwaterman) [inline src/datasources/system/components/SystemQueryEditor.tsx:72]: @MelissaHilliard - Please review this string when you get a second.

Comment (MelissaHilliard) [reply src/datasources/system/components/SystemQueryEditor.tsx:72]:
```suggestion
  workspace: `The workspace to search for the system specified.`,
```
We usually say "specified" but it's not a big deal either way

File: src/datasources/system/SystemDataSource.ts [hunk 2]
```diff
@@ -33,7 +34,11 @@ export class SystemDataSource extends DataSourceBase<SystemQuery> {
         ],
       };
     } else {
-      const metadata = await this.getSystemMetadata(this.templateSrv.replace(query.systemName, options.scopedVars));
+      const metadata = await this.getSystemMetadata(
+        this.templateSrv.replace(query.systemName, options.scopedVars),
+        defaultProjection,
+        this.templateSrv.replace(query.workspace, options.scopedVars)
+      );
```
Comment (mure) [inline src/datasources/system/SystemDataSource.ts:41]: I'm getting an error if I put in both a system and a workspace in the panel query editor. The generated filter is:

`id = "asdf" || alias = "asdf" workspace = "e2897cf7-6332-433c-8284-65b7b628f3f6"`

That's my fault - I think the join in `getSystemMetadata` needs to use `&&` as the separator. Although now that I think about it, it's not much of a use case to filter by both a specific system and a workspace. I guess you could have systems with the same name in different workspaces?

File: src/datasources/system/SystemDataSource.test.ts [hunk 3]
```diff
@@ -114,6 +114,28 @@ test('queries for system variable values - single workspace', async () => {
   expect(backendSrv.fetch.mock.lastCall?.[0].data).toHaveProperty('filter', 'workspace = "1"');
 });
 
+test('attempts to replace variables in metadata query', async () => {
+  const workspaceVariable = '$workspace';
+  backendSrv.fetch.mockReturnValue(createFetchResponse({ data: fakeSystems.map(({ id, alias }) => ({ id, alias })) }));
```
Comment (mure) [inline src/datasources/system/SystemDataSource.test.ts:119]:
```suggestion
  backendSrv.fetch.mockReturnValue(createFetchResponse({ data: fakeSystems }));
```

It doesn't really matter in this test, but this is the more correct response for the metadata query.

File: src/datasources/system/SystemDataSource.ts [hunk 4]
```diff
@@ -70,7 +75,7 @@ export class SystemDataSource extends DataSourceBase<SystemQuery> {
   }
 
   async metricFindQuery({ workspace }: SystemVariableQuery): Promise<MetricFindValue[]> {
-    const metadata = await this.getSystemMetadata('', ['id', 'alias'], workspace);
+    const metadata = await this.getSystemMetadata('', ['id', 'alias'], this.templateSrv.replace(workspace));
```
Comment (mure) [inline src/datasources/system/SystemDataSource.ts:78]: I played around with this locally. It's cool that the chained variable just works without any extra configuration.

**Review by mure — COMMENTED** [review]
Maybe we could add a test to `SystemVariableQueryEditor.test.tsx` that tests if a user can select a variable option.

**Comment by mure** [comment]
image

It would be nice if we could not add the variable you're currently editing to the dropdown, but I'm not sure how to know that.

File: src/datasources/system/SystemDataSource.ts (hunk 2 above)
Comment (joshuaprewitt) [reply src/datasources/system/SystemDataSource.ts:41]: I agree that filtering by system and workspace isn't really needed. I'm pretty sure that minionIDs have to be unique at least at the salt level. The workspace filter is mainly needed when viewing the metadata for all systems as well as for the system variable to limit the systems in the drop-down.

Comment (cameronwaterman) [reply src/datasources/system/SystemDataSource.ts:41]: Okay, I'll look into only showing the workspace drop-down when "All systems" is selected.

**Comment by cameronwaterman** [comment]
> image
> 
> It would be nice if we could not add the variable you're currently editing to the dropdown, but I'm not sure how to know that.

Yeah I thought this as well, but wasn't having much luck actually getting the current variable name in the context of the variable query editor. I'll spend a bit more time looking into this today.

File: src/core/utils.ts [hunk 5]
```diff
@@ -47,7 +47,8 @@ export function useWorkspaceOptions<DSType extends DataSourceBase<any>>(datasour
 export function getVariableOptions<DSType extends DataSourceBase<any>>(datasource: DSType) {
   return datasource.templateSrv
     .getVariables()
-    .map((variable) => ({ label: '$' + variable.name, value: '$' + variable.name }));
+    .filter(variable => (variable as any).current.value)
```
Comment (cameronwaterman) [inline src/core/utils.ts:50]: @mure - Here is my proposed solution to not showing the newly-created variable in the drop-down. It's hacky, but seems to work fairly well in my testing. Let me know what you think.

Comment (mure) [reply src/core/utils.ts:50]: Looks like query variables have a `datasource` property:

image

Maybe we could hide variables that are from the same data source the user is currently creating a variable for?

I'm also fine with not doing anything. It's not really a big deal that users see an extra autocomplete option.

File: src/test/fixtures.ts [hunk 6]
```diff
@@ -18,9 +18,26 @@ export function setupDataSource<T>(
       },
     }
   );
+  const mockVariables: TypedVariableModel[] = [{
```
Comment (mure) [inline src/test/fixtures.ts:21]: nit: I'd prefer to have this constant outside of the function for readability

**Review by mure — APPROVED** [review]

File: src/core/utils.ts (hunk 5 above)
Comment (cameronwaterman) [reply src/core/utils.ts:50]: I've gone with checking the data source, it is cleaner

**Pull request merged**